### DISCUSSION TITLE
Building and installing nanobind library with CMake

### DISCUSSION THREAD

**Discussion by yurybura** [wjakob/nanobind#341] [Ideas]
I would like to implement a [vcpkg](https://vcpkg.io/en/index.html) port for the `nanobind` library. This package manager uses CMake internally and supports many target platforms (Windows, Linux, MacOS, etc.). Using a package manager, the `nanobind` library can be built once and used in many projects without rebuilding the dependencies.

Unfortunately `nanobind` doesn't support building and installing using CMake (Pip or Conda only). This is a big limitation because the C/C++ package manager doesn't manage Python packages.

I would suggest doing an alternative way of building a static or shared library (depending on the options requested), [installing](https://cmake.org/cmake/help/latest/command/install.html) the binaries and headers, generating CMake configuration files using the [helpers](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html).

Related to https://github.com/microsoft/vcpkg/issues/33401.
Related to #55, https://github.com/wjakob/nanobind/discussions/291, https://github.com/wjakob/nanobind/issues/339.

**Comment by wjakob** [comment]
A few quick thoughts: nanobind is designed to be compiled on the fly for each project, and it's small enough that this takes literally a second. Distributing pre-compiled versions of nanobind is not something that I would endorse. (It would also be complicated, because you would have to do this for different platforms, Python versions, while also handling the various parameters that can be passed to nanobind during compilation).

In this way, nanobind is different from, say, `libpng` or other more standard libraries that are supported via headers and pre-compiled libraries. The CMake build system is part of the distribution, which is why adopting other build tools is tricky. 

Now that said, nanobind *can* be imported through CMake. In fact, this is how the pip distribution mechanism works. The PyPI file contains the CMake infrastructure and the C++ headers and source files so that other packages can run `find_package` and then have access to the build system. If that is supoprted by `vcpkg` (don't know enough about it), then it's probably not too difficult to integrate.

**Comment by ekilmer** [comment]
I have a [`refactor-cmake` branch](https://github.com/wjakob/nanobind/compare/master...ekilmer:nanobind:refactor-cmake) that I've been working on (hoping to integrate upstream, if interested/acceptable) that refactors the CMake of this project to support packaging with vcpkg and other package managers.

Fair warning: the changes in that branch aren't exactly minimal (nor all necessary), but I'm happy to break it up into smaller pieces. I think opening a PR with that branch and then have people comment, break out, document, or clean up as needed into other smaller PRs would be a decent idea? What do you think @wjakob?

Importantly, I've kept support for compiling nanobind on the fly, i.e., my refactor installs the source files of nanobind and CMake to support finding/building, like how it's done now. While I _think_ vcpkg should be fine with this, I also agree that it makes more sense to distribute nanobind this way for the reasons @wjakob mentions.

Again, I know there are quite a few changes, and I'm happy to break some of it up (or explain it) for incremental merging/review, if there's a chance for integrating and supporting this. I also tried to update the documentation (let me know if I can add anything else), but for now, the CI (and `ccmake`) should serve as documentation.

I've been testing for compatibility with three branches on the nanobind_example repo:
1. [`backward-compat`](https://github.com/wjakob/nanobind_example/compare/master...ekilmer:nanobind_example:backward-compat) branch to make sure we're backward compatible with the current code.
2. [`add_subdirectory-compat`](https://github.com/wjakob/nanobind_example/compare/master...ekilmer:nanobind_example:add_subdirectory-compat) branch to ensure nanobind can still be used with CMake's `add_subdirectory` as a submodule.
3. [`vcpkg-support`](https://github.com/wjakob/nanobind_example/compare/master...ekilmer:nanobind_example:vcpkg-support) to show an example of how to integrate nanobind as a vcpkg port, along with some other native dependencies. I also added support for building (with vcpkg) on all cibuildwheel platforms and architectures (a bit difficult and messier than expected, but not the fault of nanobind, and there is almost certainly room for improvement/simplification here)

Thank you for the great project!!

**Reply by wjakob** [reply]
My main gripe with this change is actually the point of it: it adds support for another build system.

I have experience with pure CMake -- I use it in all of my open source projects, so it's easy to keep CMake interface in perfect shape. The maintenance overheads of dealing with a single build system in what is a spare time hobby project are manageable.

But there are so many other ones: off the to of my head, there are

- meson
- xmake
- vcpkg
- premake
- bazel
- gradle

And probably even more that I missed. There is just no way. If users wish to have nanobind interoperability with these tools, they will have to roll their own build system and evolve in sync with the development of nanobind (without me having to think about it.)

A good example of this sort of approach for pybind11 is shown here: https://github.com/pybind/pybind11_bazel

"vcpkg" is a bit special since it at least seems to piggyback on top of CMake.  But I would not be able to ensure that I'm not breaking things all the time. (It's not so easy, as demonstrated by the large size of the diff).

**Reply by wjakob** [reply]
I just committed a FAQ entry that outlines the rationale a bit more, since this question is likely to come up again in the future.

**Reply by yurybura** [reply]
`vcpkg` is not a build system, but a package manager. All that is needed is support for the standard procedure for building and installing the library using CMake (named pure CMake). C/C++ package managers, such as `vcpkg` or `conan`, already have recipes for handling CMake-based libraries.

Your wonderful project is already using CMake. This requires at least one external dependency `robin_map`, the source code of which is not included in the [zip](https://github.com/wjakob/nanobind/archive/refs/tags/v1.8.0.zip) or [tar.gz](https://github.com/wjakob/nanobind/archive/refs/tags/v1.8.0.tar.gz). It is not enough to simply download release version 1.8.0 (or 1.7.0) from Github and run the CMake install command. Every Python environment for every project on my PC needs to be modified if it uses `nanobind`.
 
The package manager is very easy to use for end users. It supports caching which helps avoid unnecessary downloads, builds and installations.

**Reply by ekilmer** [reply]
> My main gripe with this change is actually the point of it: it adds support for another build system.

Not to be pedantic, but vcpkg is a _packaging_ solution, like RPM, `.deb`, `PKGBUILD`, etc. rather than a build system like the others you've listed. Vcpkg supports packaging projects with the majority of build tools (popular with C++ developers) that you've listed, in addition to CMake. 

The only thing that's missing from the current nanobind solution are CMake `install` commands. If these were added, then every packaging solution that understands CMake projects could distribute nanobind. My branch adds these `install` commands, albeit with quite a few more changes to follow some standard best practices that I've picked up during my experiences with CMake project creation (certainly, these are subjective).

Also, my branch doesn't change any of the user-facing CMake interface, and if there are breaking changes, that was not my intention and should be fixed. I had hoped this was evidenced through the two compatibility branches I linked in my fork of `nanobind_example`: [`backward-compat`](https://github.com/wjakob/nanobind_example/compare/master...ekilmer:nanobind_example:backward-compat) and [`add_subdirectory-compat`](https://github.com/wjakob/nanobind_example/compare/master...ekilmer:nanobind_example:add_subdirectory-compat). I had also included the vcpkg branch to show how nanobind would be packaged in vcpkg and then used in a project with other native dependencies that cibuildwheel could build.

I hope I am not missing addressing some other big concern in your response. Thank you for your time and thoughts.

**Reply by wjakob** [reply]
Apologies, it seems that I misunderstood the purpose of `vcpkg` then. There is one basic thing I don't understand though: nanobind is a project in the Python ecosystem, where the two main delivery mechanisms for dependencies are Pip and Conda. What would be a reason to involve yet another package manager? It seems to me that the user will in any case need to install other PyPI or Conda packages to build any nontrivial extension, at which point fetching nanobind that way is easy without requiring another tool?

Now, just to be clear about the high-level idea: I am not at all opposed to adding build targets—there was another request for this a while back, and it seems reasonable. But this PR is very big, and I think I actually disagree with many changes that aren't related to mere addition of `install(..)` targets. Already a few thoughts:

- one change removes all of the nanobind headers from the `add_library` call, which seems like it would interfere with the tree-based project structure in an IDE like Visual Studio (classic) or XCode.
- The changes to `setup.py` make this file far more complicated than it seems it needs to be. 
- The changes to the main `CMakeLists.txt` file are very long (and mostly additions, not changes/removals). What's the absolute minimum version of this that suffices to add install targets?
- I also didn't understand why `nanobind-config.cmake` was renamed, and why the `find_package` at the top was wrapped into a macro.
- The added testing infrastructure is far too heavy to me given the free tier GHA status of this project, especially adding tests with respect to all sorts of array frameworks. I do test these regularly, just not on GHA.
- I also would prefer vcpkg-related testing infrastructure and `.json` files to be kept out of the repository.

**Reply by yurybura** [reply]
> Apologies, it seems that I misunderstood the purpose of `vcpkg` then. There is one basic thing I don't understand though: nanobind is a project in the Python ecosystem, where the two main delivery mechanisms for dependencies are Pip and Conda. What would be a reason to involve yet another package manager? It seems to me that the user will in any case need to install other PyPI or Conda packages to build any nontrivial extension, at which point fetching nanobind that way is easy without requiring another tool?

If I understand correctly, nanobind is a fairly young and promising library that is useful for C++ developers.

This is my use case:
* I am the developer of a C++ project/library.
* This C++ project uses CMake and has many external dependencies (GTest, ZLib, ZStd, LZ4, {fmt}, Magic-enum, Boost, nanobind [planned], etc.). External dependencies are handled in CMake (find_package, find_dpendency, etc.). vcpkg is used to install these dependencies on the build machine.
* This C++ project provides released pre-built binaries (for some reason, such as the build procedure being too complex or too long).
* This C++ project requires Python bindings. Bindings are also released as pre-built binaries (many .pyd files).
* End users of this project download/install the pre-build binaries and use them with various versions of Python.

Pre-build binaries must be provided for all supported versions of Python and the Stable ABI. nanobind sources is only needed when building binaries. There will be many variations of pre-built binaries related to Python version (3.8...3.12), OS, CPU architecture.
I would like to build nonobind once as a shared library and use it in every .pyd module. To make dependency handling easier I want to still use vcpkg.

Finally, the nanobind port for vcpkg (the nanobind installation recipe) will be committed to the vcpkg repository. Of course, it is possible to make a patch in the vcpkg repository to support the CMake install command, but it is much better to have it here.

**Reply by ekilmer** [reply]
TL;DR, I got carried away with CMake changes, and I will create a more focused branch with the minimal requirements for packaging with vcpkg---likely just the addition of CMake `install` commands.

----------------

> nanobind is a project in the Python ecosystem, where the two main delivery mechanisms for dependencies are Pip and Conda. What would be a reason to involve yet another package manager?

Along with what @yurybura has said, the reason would be for flexibility and freedom in how a user is able to declare and specify their dependencies. If a user is already getting all of their other C++ dependencies from vcpkg or _wherever_, then they would also likely want to get nanobind from there.

> It seems to me that the user will in any case need to install other PyPI or Conda packages to build any nontrivial extension, at which point fetching nanobind that way is easy without requiring another tool?

This may be true, and the option to install/use with PyPI or Conda would still be available!

> one change removes all of the nanobind headers from the add_library call, which seems like it would interfere with the tree-based project structure in an IDE like Visual Studio (classic) or XCode.

Good point. I will add these back, but I think this used to be a problem in the past, and IDEs have improved, but I'd have to experiment. Explicitly listing headers here might be necessary because they're outside of the user's CMake source directory, and IDEs probably wouldn't expect that to be the case.

> The changes to setup.py make this file far more complicated than it seems it needs to be.

Yeah, I don't like how this turned out either, and it only became this complicated due to support for correct `sdist` handling. I thought that maybe `scikit-build-core` would take care of it better, but I wasn't sure how to dynamically parse the version string from a C++ header file. I need to read up more on modern Python packaging and how this might be possible.

I also inserted these changes so that CMake is the only place that controls how nanobind is installed (even with Python packaging). I was hoping to deduplicate installation logic.

> The changes to the main CMakeLists.txt file are very long (and mostly additions, not changes/removals). What's the absolute minimum version of this that suffices to add install targets?

Sorry about that, I got carried away with experimenting. I should have presented a minimal version when I first posted here. I will do this.

> I also didn't understand why `nanobind-config.cmake` was renamed

Ideally, nanobind's CMake installation rules would allow the user/packager to control (without making a patch) where certain directories are installed, i.e. source (`/src`), header (`/include/nanobind`), and data files (`/share/nanobind`) (including CMake (`/share/nanobind/cmake`)). To support this, we need to generate a `nanobind-config.cmake` file (from `nanobind-config.cmake.in`) with the information passed by the user during CMake configuration.

Certain packaging ecosystems have different policies/standards around the specific locations and directory hierarchy of how a package is installed. The biggest example is where CMake files are installed: is it `/share/nanobind/cmake` or `/share/cmake/nanobind` or something different? The easiest way to support all scenarios is by providing cache variables to let the user set their own option and then persist that option into a generated `nanobind-config.cmake`.

However, thinking more, this change would break any user who tried to `find_package(nanobind)` and set `nanobind_DIR` to the directory containing `nanobind-config.cmake` in nanobind's repository directory. Was this an intended use-case of nanobind? If so, that might be difficult to continue supporting without some duplication of logic.

> and why the `find_package` at the top was wrapped into a macro.

We need to call this in two places, so I didn't want to repeat myself. The first call is directly after the macro definition, because we need use the info from CMake's Python module for 

I encountered some issues with target scoping when using nanobind with `add_subdirectory`. Ideally, the user should not have to specify nanobind's dependencies in their own CMake---nanobind should find its own dependencies. However, this is an issue because CMake's `IMPORTED` targets are scoped to a directory, and so when the user does `add_subdirectory(nanobind)`, the `IMPORTED` targets are not available when the user calls the provided functions that require those targets during linking.

The reason for a [`macro` vs `function`](https://cmake.org/cmake/help/latest/command/macro.html#macro-vs-function) is because I didn't want to worry about more CMake scoping issues:

    macro is executed as if the macro body were pasted in place of the calling statement.

> The added testing infrastructure is far too heavy to me given the free tier GHA status of this project, especially adding tests with respect to all sorts of array frameworks. I do test these regularly, just not on GHA.

Understood.

> I also would prefer vcpkg-related testing infrastructure and .json files to be kept out of the repository.

Understood.

**Reply by ekilmer** [reply]
Okay, I've created a PR #356 for review with minimal changes required for packaging nanobind with CMake installation rules.

Aside from nanobind's CI, I've also created another 3 branches in my fork of `nanobind_example` to increase confidence that the installation rules work as desired and do not break compatibility:

1. [vcpkg-minimal-nanobind](https://github.com/ekilmer/nanobind_example/tree/vcpkg-minimal-nanobind) has (hopefully) what would be included upstream in vcpkg in [`.vcpkg/ports/nanobind`](https://github.com/ekilmer/nanobind_example/tree/vcpkg-minimal-nanobind/.vcpkg/ports/nanobind). The rest of the changes in that branch probably aren't relevant for the nanobind repo, but it serves as an example for a project using nanobind and vcpkg with passing CI
2. [backward-compat-minimal](https://github.com/ekilmer/nanobind_example/tree/backward-compat-minimal) shows that CI passes with my branch using a pip-installed version of nanobind.
3. [add_subdirectory-minimal-compat](https://github.com/ekilmer/nanobind_example/tree/add_subdirectory-minimal-compat) shows that CI passes with my branch when including nanobind with `add_subdirectory`

**Reply by wjakob** [reply]
I was curious to see if a `vcpkg` package for nanobind now exists and couldn't find one. Are there remaining issues to be addressed?

**Reply by yurybura** [reply]
[The PR](https://github.com/microsoft/vcpkg/pull/31323) is exist but haven't done yet. It should be tested with a new `nanobind` release.

**Reply by ekilmer** [reply]
I also just created a PR https://github.com/microsoft/vcpkg/pull/35488. I need to investigate the failing Windows builds. Seems to only be when Python is built dynamically. It's probably my fault.

**Reply by ekilmer** [reply]
The vcpkg PR (https://github.com/microsoft/vcpkg/pull/35488) CI is now passing, after adding the `Py_DEBUG` compilation flag to the debug builds for the CI test example package. I also added an allowed failure for Windows arm64 cross-compilation because there were issues with finding/running the Python executable.

I'll request review in vcpkg after a new nanobind release is made. I don't believe any new large changes will be needed, and if there are, then they can be incorporated as a patch in vcpkg and upstreamed into a new release at a later date.

**Reply by wjakob** [reply]
It will likely take a long while for the next release. Is there a particular reason to not proceed with the current one?
